### ISSUE TITLE
Arrow function with wrong return type relative to explicit signature does not cause compiler errors.

### ISSUE THREAD

**Issue by brunnerh** [Microsoft/TypeScript#21674]
**TypeScript Version:**  2.8.0-dev.20180206

**Search Terms:**
wrong return type arrow function
inferred type arrow function

**Code**

```ts
// Does not compile, because the signature does not match the inferred return type.
function test1(value: number): void
{
	return value; // error TS2322: Type 'number' is not assignable to type 'void'.
}

// Does compile, type is not correctly inferred, signature is used.
const test2: (value: number) => void
	= (value: number) => value;

test2(42) + 1; // error TS2365: Operator '+' cannot be applied to types 'void' and '1'.
```

**Expected behavior:**

`test2` fails to compile because the signature does not match the returned object's type, just as `test1` does.

**Actual behavior:**

`test2` compiles and uses the explicit signature without error.

**Playground Link:** [[Link]](https://www.typescriptlang.org/play/#src=%2F%2F%20Does%20not%20compile%2C%20because%20the%20signature%20does%20not%20match%20the%20inferred%20return%20type.%0D%0Afunction%20test1(value%3A%20number)%3A%20void%0D%0A%7B%0D%0A%09return%20value%3B%20%2F%2F%20error%20TS2322%3A%20Type%20'number'%20is%20not%20assignable%20to%20type%20'void'.%0D%0A%7D%0D%0A%0D%0A%2F%2F%20Does%20compile%2C%20type%20is%20not%20correctly%20inferred%2C%20signature%20is%20used.%0D%0Aconst%20test2%3A%20(value%3A%20number)%20%3D%3E%20void%0D%0A%09%3D%20(value%3A%20number)%20%3D%3E%20value%3B%0D%0A%0D%0Atest2(42)%20%2B%201%3B%20%2F%2F%20error%20TS2365%3A%20Operator%20'%2B'%20cannot%20be%20applied%20to%20types%20'void'%20and%20'1'.)

**Related Issues:**
#17505 - About argument inference, rather than return type.

**Comment by mhegazy** [comment]
The first case is clearly a bug. the compiler alerts you that the declared API does not match your implementation..

The second is different, u are allowed to assign something that is more specific to something that is more general.. for instance, `var x: Animal =  new Dog()` is a valid assignment.. similarly, you can assign something that returns a value to one that does not require a value.. the error is correct, since `test2` is declared to return a `void`. that is the same as `const x: Animal = new Dog(); x.bark()`

**Comment by RyanCavanaugh** [comment]
https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-functions-returning-non-void-assignable-to-function-returning-void

**Comment by brunnerh** [comment]
I understand the reasoning here but i slightly disagree about the statement `This is the expected and desired behavior.`

To take the example from the FAQ:
```ts
function doSomething(): number {
    return 42;
}
function callMeMaybe(callback: () => void) {
    callback();
}
function
callMeMaybe(doSomething);
```

This does absolutely nothing, since `doSomething` is a pure function and `callMeMaybe` ostensibly expects to be passed a function with side effects, as that is the only thing a function with such a signature could achieve.

This feature of allowing that assignment has just brilliantly masked a coding error that could have been avoided.

Personally, i think another strict flag would be great to make this sort of assignment illegal. To reasonably work with such a flag, something like an [ignore function](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/operators.ignore%5B't%5D-function-%5Bfsharp%5D) would be great, so you can make the second example work without much hassle:

```ts
let items = [1, 2];
callMeMaybe(() => ignore(items.push(3)));
```

**Comment by RyanCavanaugh** [comment]
This has been discussed countless times, see #19014, #8240, #8584, #8581, #9603, #8615, etc
